File: go/test/custom_field_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/shortcut-sdk/go"
	"github.com/voxgig-sdk/shortcut-sdk/go/core"

	vs "github.com/voxgig-sdk/shortcut-sdk/go/utility/struct"
)

func TestCustomFieldEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.CustomField(nil)
		if ent == nil {
			t.Fatal("expected non-nil CustomFieldEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"custom_field": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.CustomField(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.CustomField(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := custom_fieldBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "custom_field." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set SHORTCUT_TEST_CUSTOM_FIELD_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		customFieldRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.custom_field", setup.data)))
		var customFieldRef01Data map[string]any
		if len(customFieldRef01DataRaw) > 0 {
			customFieldRef01Data = core.ToMapAny(customFieldRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = customFieldRef01Data

		// LIST
		customFieldRef01Ent := client.CustomField(nil)
		customFieldRef01Match := map[string]any{}

		customFieldRef01ListResult, err := customFieldRef01Ent.List(customFieldRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, customFieldRef01ListOk := customFieldRef01ListResult.([]any)
		if !customFieldRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", customFieldRef01ListResult)
		}

		// UPDATE
		customFieldRef01DataUp0Up := map[string]any{
			"id": customFieldRef01Data["id"],
		}

		customFieldRef01MarkdefUp0Name := "after_id"
		customFieldRef01MarkdefUp0Value := fmt.Sprintf("Mark01-custom_field_ref01_%d", setup.now)
		customFieldRef01DataUp0Up[customFieldRef01MarkdefUp0Name] = customFieldRef01MarkdefUp0Value

		customFieldRef01ResdataUp0Result, err := customFieldRef01Ent.Update(customFieldRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		customFieldRef01ResdataUp0 := core.ToMapAny(entityData(customFieldRef01ResdataUp0Result))
		if customFieldRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if customFieldRef01ResdataUp0["id"] != customFieldRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if customFieldRef01ResdataUp0[customFieldRef01MarkdefUp0Name] != customFieldRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", customFieldRef01MarkdefUp0Name, customFieldRef01ResdataUp0[customFieldRef01MarkdefUp0Name])
		}

		// LOAD
		customFieldRef01MatchDt0 := map[string]any{
			"id": customFieldRef01Data["id"],
		}
		customFieldRef01DataDt0Loaded, err := customFieldRef01Ent.Load(customFieldRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		customFieldRef01DataDt0LoadResult := core.ToMapAny(entityData(customFieldRef01DataDt0Loaded))
		if customFieldRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if customFieldRef01DataDt0LoadResult["id"] != customFieldRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func custom_fieldBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "custom_field", "CustomFieldTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read custom_field test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse custom_field test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"custom_field01", "custom_field02", "custom_field03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("SHORTCUT_TEST_CUSTOM_FIELD_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"SHORTCUT_TEST_CUSTOM_FIELD_ENTID": idmap,
		"SHORTCUT_TEST_LIVE":      "FALSE",
		"SHORTCUT_TEST_EXPLAIN":   "FALSE",
		"SHORTCUT_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["SHORTCUT_TEST_CUSTOM_FIELD_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["SHORTCUT_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["SHORTCUT_APIKEY"],
			},
			extra,
		})
		client = sdk.NewShortcutSDK(core.ToMapAny(mergedOpts))
	}

	live := env["SHORTCUT_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["SHORTCUT_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/custom_field_entity_test.rb
# CustomField entity test

require "minitest/autorun"
require "json"
require_relative "../Shortcut_sdk"
require_relative "runner"

class CustomFieldEntityTest < Minitest::Test
  def test_create_instance
    testsdk = ShortcutSDK.test(nil, nil)
    ent = testsdk.CustomField(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "custom_field" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = ShortcutSDK.test(seed, nil)
    seen = base.CustomField(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = ShortcutConfig.shared_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = ShortcutSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.CustomField(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = custom_field_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list", "update", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "custom_field." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set SHORTCUT_TEST_CUSTOM_FIELD_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    custom_field_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.custom_field")))
    custom_field_ref01_data = nil
    if custom_field_ref01_data_raw.length > 0
      custom_field_ref01_data = Helpers.to_map(custom_field_ref01_data_raw[0][1])
    end

    # LIST
    custom_field_ref01_ent = client.CustomField(nil)
    custom_field_ref01_match = {}

    custom_field_ref01_list_result = custom_field_ref01_ent.list(custom_field_ref01_match, nil)
    assert custom_field_ref01_list_result.is_a?(Array)

    # UPDATE
    custom_field_ref01_data_up0_up = {
      "id" => custom_field_ref01_data["id"],
    }

    custom_field_ref01_markdef_up0_name = "after_id"
    custom_field_ref01_markdef_up0_value = "Mark01-custom_field_ref01_#{setup[:now]}"
    custom_field_ref01_data_up0_up[custom_field_ref01_markdef_up0_name] = custom_field_ref01_markdef_up0_value

    custom_field_ref01_resdata_up0_result = custom_field_ref01_ent.update(custom_field_ref01_data_up0_up, nil)
    custom_field_ref01_resdata_up0 = Helpers.to_map(custom_field_ref01_resdata_up0_result.respond_to?(:data_get) ? custom_field_ref01_resdata_up0_result.data_get : custom_field_ref01_resdata_up0_result)
    assert !custom_field_ref01_resdata_up0.nil?
    assert_equal custom_field_ref01_resdata_up0["id"], custom_field_ref01_data_up0_up["id"]
    assert_equal custom_field_ref01_resdata_up0[custom_field_ref01_markdef_up0_name], custom_field_ref01_markdef_up0_value

    # LOAD
    custom_field_ref01_match_dt0 = {
      "id" => custom_field_ref01_data["id"],
    }
    custom_field_ref01_data_dt0_loaded = custom_field_ref01_ent.load(custom_field_ref01_match_dt0, nil)
    custom_field_ref01_data_dt0_load_result = Helpers.to_map(custom_field_ref01_data_dt0_loaded.respond_to?(:data_get) ? custom_field_ref01_data_dt0_loaded.data_get : custom_field_ref01_data_dt0_loaded)
    assert !custom_field_ref01_data_dt0_load_result.nil?
    assert_equal custom_field_ref01_data_dt0_load_result["id"], custom_field_ref01_data["id"]

  end
end

def custom_field_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "custom_field", "CustomFieldTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = ShortcutSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["custom_field01", "custom_field02", "custom_field03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["SHORTCUT_TEST_CUSTOM_FIELD_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "SHORTCUT_TEST_CUSTOM_FIELD_ENTID" => idmap,
    "SHORTCUT_TEST_LIVE" => "FALSE",
    "SHORTCUT_TEST_EXPLAIN" => "FALSE",
    "SHORTCUT_APIKEY" => "",
  })

  idmap_resolved = Helpers.to_map(
    env["SHORTCUT_TEST_CUSTOM_FIELD_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["SHORTCUT_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      # FIRST, so the generated fields below win: sdk-test-control.json's
      # test.client.options adds to the live client, it does not redirect it.
      Runner.live_client_options,
      {
        "apikey" => env["SHORTCUT_APIKEY"],
      },
      extra || {},
    ])
    client = ShortcutSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["SHORTCUT_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["SHORTCUT_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
